### DISCUSSION TITLE
Change the way Termux packages are updated

### DISCUSSION THREAD

**Discussion by amygrinn** [termux/termux-packages#19616] [Q&A] [answered]
The vulnerability discovered with liblzma led to me to realize how bleeding edge termux is. Is there another update channel for Termux which is more stable?

**Comment by twaik** [comment]
No. Termux has only one update channel.

**Reply by amygrinn** [reply]
Since there is only one update channel, should the packages be updated as frequently as they currently are?

**Reply by twaik** [reply]
Yes. We update packages as soon as possible and if there are no porting problems.

**Comment by agnostic-apollo** [comment, marked as answer]
Termux officially only provides one channel and uses a rolling release model. We do not have enough maintainers to resolve issues with compatibility with multiple versions of dependencies. Moreover, our server also has limited space, so we cannot keep multiple versions/channels, although cost of space is not an issue these days like it used to be a few years ago when we were relying on donated servers.

Lot of users want latest features provided by softwares, so we generally update packages whenever upstream releases an update.

Having a `stable` while possible, would require more maintenance effort.

**Reply by amygrinn** [reply]
I understand that there can only be one channel at the moment, but I believe that concern for users preferences and security should be balanced. There is obviously some risk involved no matter how packages are released, but the way it's being done now seems a little dangerous. It also appears to me to be *more* maintenance work to have the latest version supported, whereas primarily accepting bugfixes would be less likely to require manual intervention. Can you explain more the decision to build the newest releases?

**Reply by agnostic-apollo** [reply]
Check https://unix.stackexchange.com/a/112160 for what problems would arise if we only provided some old/stable version of a dependency library. Users wouldn't be able install latest versions of a package that requires it to be compiled against a higher version of the library. Since newer version wouldn't be installable and they may have bug or security fixes, we would then have to backport those ourselves to some older version of the package that is compatible with old version of a library, we do not have resources for that. It's easier to just keep all packages to their latest version and recompile them whenever a library is updated.

See also
- https://www.debian.org/doc/debian-policy/ch-sharedlibs.html#shared-library-abi-changes
- https://unix.stackexchange.com/a/289551
- https://github.com/termux/termux-packages/issues/8649

**Reply by agnostic-apollo** [reply]
>It also appears to me to be more maintenance work to have the latest version supported

Maintainers here generally jump bump the version of the package when its update is released in upstream, and that's mostly automated now. If it compiles and patches are still valid, then it is pushed to our repo. If it fails to compile, then its looked into why. Only sometimes manual testing is done for specific updates, and upstream changes are rarely checked, unless a major upgrade, as we only have 5-10 active package maintainers who are updating over 2000+ packages, its not physically possible for them to check changes in upstream for every release of every package. Only when compilation or runtime errors occur, only then things are looked at mostly.

**Reply by TomJo2000** [reply]
Shipping out of date software does not protect against vulnerabilities in those versions either.
As Apollo has gone over in detail, an LTS model just does not work for us.

**Reply by twaik** [reply]
> Only when compilation or runtime errors occur, only then things are looked at mostly.

Actuqlly when github will release aarch64 runner we may try to write some termux-rootfs runtime consistency check to ensure that if we update some library all reverse-dependencies can be started and run self-checks (or special tests we can write or package tests provided by packages). So we will be sure package update does not brake basic stuff.

**Reply by agnostic-apollo** [reply]
>Shipping out of date software does not protect against vulnerabilities in those versions either

Unless backports are done like linux distros do, which we don't have resources for. We actually do not even have the time to review every package to see if there are bug or security fixes in newer versions, let alone do backports for them.

>runner we may try to write some termux-rootfs runtime consistency

Yeah, such testing is of course planned, but currently not done.

**Reply by amygrinn** [reply]
I understand that most package upgrades are necessarily unattended, my question is more:

If version 2.3.1 of some package is currently available in Termux and the upstream maintainer releases 2.3.2 and 2.4.0, which version should be chosen?

To me the former would be less likely to result in manual intervention, while the latter is likely to introduce new features that may interact in unknown ways. I don't see a compelling reason that liblzma needed to be upgraded in February from the discussion and links shared here. Nor do I see how upgrading less often would necessitate back ports or increase maintenance costs.

I'm suggesting that minor package upgrades are released on a more as-needed basis rather than being pushed directly from upstream.

**Reply by twaik** [reply]
> If version 2.3.1 of some package is currently available in Termux and the upstream maintainer releases 2.3.2 and 2.4.0, which version should be chosen?

If 2.4.0 version of package does not break compatibility with other packages (it is critical in the case if it is a library) and there are no porting problems we will choose 2.4.0.

**Reply by agnostic-apollo** [reply]
>If version 2.3.1 of some package is currently available in Termux and the upstream maintainer releases 2.3.2 and 2.4.0, which version should be chosen?

As twaik said, we would likely go with `2.4.0` as that will have the latest features, unless there are porting problems.

>while the latter is likely to introduce new features that may interact in unknown ways.

That is just the risk of a rolling release model. Check unstable/testing channels for debian.

- https://wiki.debian.org/DebianUnstable
- https://wiki.debian.org/DebianTesting

> I don't see a compelling reason that liblzma needed to be upgraded in February from the discussion and links shared here. Nor do I see how upgrading less often would necessitate back ports or increase maintenance costs.

How do you suppose we find compelling reason for whether a package should be upgraded or not? That would require looking at changelogs and commits and possibly related issues/pulls of the changes to decide if its worth it. Not all packages would strictly follow semantic versioning design for changes. We do not have the resources to check for such changes for updates of every package.

>Nor do I see how upgrading less often would necessitate back ports or increase maintenance costs.

Of course it would require us to work on backports if we cannot upgrade to latest version, just as debian distro does it, as upstreams don't always provide backports themselves, and its only done if they have some LTS release design themselves...

- https://backports.debian.org/Instructions/

>I'm suggesting that minor package upgrades are released on a more as-needed basis rather than being pushed directly from upstream.

That wouldn't necessarily protect from malicious code, it could just as easily have been added to a major upgrade too and not found out as it was very obfuscated. `xz-utils` releases are manually merged into debian forks, but malicious changes weren't caught by their maintainers either. There were thousands of lines of changes in updates too.

- https://salsa.debian.org/debian/xz-utils/-/commit/12388833e66a4ddafe08571882ad638a511cf68b
- https://salsa.debian.org/debian/xz-utils/-/commit/f7c276c144009b4b6ced966a69b7dd2ae3a4b52e
